components/gfx/display_list/mod.rs

```diff
@@ -47,6 +47,75 @@ pub struct DisplayList {
     pub list: Vec<DisplayItem>,
 }
 
+struct ScrollOffsetLookup<'a> {
+    parents: &'a mut HashMap<ClipId, ClipId>,
+    calculated_total_offsets: ScrollOffsetMap,
+    raw_offsets: &'a ScrollOffsetMap,
+}
+
+impl<'a> ScrollOffsetLookup<'a> {
+    fn new(parents: &'a mut HashMap<ClipId, ClipId>,
+           raw_offsets: &'a ScrollOffsetMap)
+           -> ScrollOffsetLookup<'a> {
+        ScrollOffsetLookup {
+            parents: parents,
+            calculated_total_offsets: HashMap::new(),
+            raw_offsets: raw_offsets,
+        }
+    }
+
+    fn new_for_reference_frame(&mut self,
+                               clip_id: ClipId,
+                               transform: &Matrix4D<f32>,
+                               point: &mut Point2D<Au>)
+                               -> Option<ScrollOffsetLookup> {
+        // If a transform function causes the current transformation matrix of an object
+        // to be non-invertible, the object and its content do not get displayed.
+        let inv_transform = match transform.inverse() {
+            Some(transform) => transform,
+            None => return None,
+        };
+
+        let scroll_offset = self.full_offset_for_scroll_root(&clip_id);
+        *point = Point2D::new(point.x - Au::from_f32_px(scroll_offset.x),
+                              point.y - Au::from_f32_px(scroll_offset.y));
+        let frac_point = inv_transform.transform_point(&Point2D::new(point.x.to_f32_px(),
+                                                                     point.y.to_f32_px()));
+        *point = Point2D::new(Au::from_f32_px(frac_point.x), Au::from_f32_px(frac_point.y));
+
+        let mut sublookup = ScrollOffsetLookup {
+            parents: &mut self.parents,
+            calculated_total_offsets: HashMap::new(),
+            raw_offsets: self.raw_offsets,
+        };
+        sublookup.calculated_total_offsets.insert(clip_id, Point2D::zero());
+        Some(sublookup)
+    }
+
+    fn add_scroll_root(&mut self, scroll_root: &ScrollRoot) {
+        self.parents.insert(scroll_root.id, scroll_root.parent_id);
+    }
+
+    fn full_offset_for_scroll_root(&mut self, id: &ClipId) -> Point2D<f32> {
+        match self.calculated_total_offsets.get(id) {
```

Comment:
I cannot use the entry API here, because I'm recursing causing a double mutable borrow. `if let` is a lot cleaner though.